src/notebooks/helpers.ts

```diff
@@ -936,7 +936,16 @@ export function updateNotebookMetadata(
                 break;
         }
 
-        if (metadata.kernelspec?.name !== name) {
+        // Pull our old and new interpreter hash, if these don't match we also want to update metadata
+        // since name might be python3 in both scenarios and they might have the same python version so the
+        // check above with language info would not see them as changed.
```

Comment:
https://github.com/microsoft/vscode-jupyter/blob/a912533e456111d267f6d3838936e610387bbc03/src/notebooks/helpers.ts#L898

You can see in this line linked above that "changed" would get set to true already if python language versions are different. If they were not different you could have a comparison of 'python3' to 'python3' for the name values when the hashes were different. This is why the issue showed best if you try to swap between two python envs with the same version number.